zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,924 @@
+        input: Option<(u32, transparent::Output)>,
+    ) -> Self {
+        let input = if let Some((index, prevout)) = input {
+            let index = index as usize;
+            let inputs = trans.inputs();
+
+            Some((prevout, &inputs[index], index))
+        } else {
+            None
+        };
+
+        SigHasher {
+            trans,
+            hash_type,
+            network_upgrade,
+            input,
+        }
+    }
+
+    pub fn sighash(self) -> Hash {
+        use NetworkUpgrade::*;
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(&self.personal())
+            .to_state();
+
+        match self.network_upgrade {
+            Genesis | BeforeOverwinter => unreachable!("Zebra checkpoints on Sapling activation"),
+            Overwinter => self
+                .hash_sighash_zip143(&mut hash)
+                .expect("serialization into hasher never fails"),
+            Sapling | Blossom | Heartwood | Canopy => self
+                .hash_sighash_zip243(&mut hash)
+                .expect("serialization into hasher never fails"),
+        }
+
+        hash.finalize()
+    }
+
+    fn consensus_branch_id(&self) -> ConsensusBranchId {
+        self.network_upgrade
+            .branch_id()
+            .expect("Zebra checkpoints on Sapling activation")
+    }
+
+    /// Sighash implementation for the overwinter consensus branch
+    fn hash_sighash_zip143<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        self.hash_header(&mut writer)?;
+        self.hash_groupid(&mut writer)?;
+        self.hash_prevouts(&mut writer)?;
+        self.hash_sequence(&mut writer)?;
+        self.hash_outputs(&mut writer)?;
+        self.hash_joinsplits(&mut writer)?;
+        self.hash_lock_time(&mut writer)?;
+        self.hash_expiry_height(&mut writer)?;
+        self.hash_hash_type(&mut writer)?;
+        self.hash_input(&mut writer)?;
+
+        Ok(())
+    }
+
+    /// Sighash implementation for the sapling consensus branch and every
+    /// subsequent consensus branch
+    fn hash_sighash_zip243<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        self.hash_header(&mut writer)?;
+        self.hash_groupid(&mut writer)?;
+        self.hash_prevouts(&mut writer)?;
+        self.hash_sequence(&mut writer)?;
+        self.hash_outputs(&mut writer)?;
+        self.hash_joinsplits(&mut writer)?;
+        self.hash_shielded_spends(&mut writer)?;
+        self.hash_shielded_outputs(&mut writer)?;
+        self.hash_lock_time(&mut writer)?;
+        self.hash_expiry_height(&mut writer)?;
+        self.hash_value_balance(&mut writer)?;
+        self.hash_hash_type(&mut writer)?;
+        self.hash_input(&mut writer)?;
+
+        Ok(())
+    }
+
+    pub(super) fn personal(&self) -> [u8; 16] {
+        let mut personal = [0; 16];
+        (&mut personal[..12]).copy_from_slice(ZCASH_SIGHASH_PERSONALIZATION_PREFIX);
+        (&mut personal[12..])
```

Comment:
Cute!

Reply:
stole this straight from @str4d's impl I think :D

Reply:
or maybe something @dconnolly wrote somewhere else, not sure, definitely stolen tho

Reply:
I have definitely done something like this elsewhere 😅